### ISSUE TITLE
cabal ignores the 'cabal-version' field when solving for setup dependencies

### ISSUE THREAD

**Issue by grayjay** [haskell/cabal#3751]
This can cause the build to fail when the `setup-depends` range for Cabal is less restrictive than the `cabal-version` for a given package. See https://github.com/haskell/cabal/pull/3723#issuecomment-243337073.

The current behavior is technically correct, because it is possible for a setup script to depend on one version of Cabal and implement the command interface from a different version.  I think it would be better to prevent build failures in the more common case, though.

One solution is to update the `setup-depends` constraint on Cabal to reflect `cabal-version` before dependency solving.

**Comment by 23Skidoo** [comment]
> This can cause the build to fail when the `setup-depends` range for Cabal is less restrictive than the `cabal-version` for a given package.

Can't we just make this an error in `D.PD.Check`?

**Comment by grayjay** [comment]
Yeah, that would also work.  Is it okay if existing packages fail the check?

**Comment by 23Skidoo** [comment]
I think that we already have some (obsolete) packages on Hackage that fail `cabal check`.

**Comment by ezyang** [comment]
@grayjay Are you planning to fix this bug? If so assign yourself and milestone this to 2.0 ;)

**Comment by grayjay** [comment]
I wasn't planning on fixing this soon.

**Comment by ezyang** [comment]
OK, I milestoned this appropriately.

**Comment by andreasabel** [comment]
Here is a recent case of this in the wild:
- https://github.com/gtk2hs/gtk2hs/issues/318

`cairo` specifies `cabal-version: 3.0`, but in the custom setup only `Cabal >= 2.0`, so a recent `cabal-install >= 3` can parse the cabal file, but the generated setup command build with `Cabal < 3` then crashes as it does not understand this cabal file format.

So, `cabal check` should check that the constraint on `Cabal` for the setup should be at least `cabal-version`.

@grayjay wrote:
> It might be better to add an implicit constraint

I also think that an implicit constraint is in order; everything else is counterintuitive.  `cabal check` should still warn.

**Comment by sellout** [comment]
This is the blurb I have in my Cabal file template, in case anyone finds it useful or wants to improve it:
``` cabal
cabal-version: 3.0

[…]

custom-setup
  setup-depends:
    -- TODO: Due to haskell/cabal#3751, `Cabal` has to be specified even though
    --       there’s no direct dependency. Its lower bound needs to match
    --      `cabal-version` at the top of this file, and Hackage requires it to
    --       have _some_ upper bound. (Since there’s no direct dependency, it
    --       doesn’t use PVP bounds.)
    Cabal >= 3.0 && < 99,

[…]
```
Even once this gets fixed, this blurb will have to stick around until the fixed version is the lower bound.
